File: jar-common/src/main/java/com/beancounter/common/contracts/PriceRequest.java
package com.beancounter.common.contracts;

import com.beancounter.common.input.AssetInput;
import com.beancounter.common.model.Asset;
import java.util.ArrayList;
import java.util.Collection;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.RequiredArgsConstructor;

@Data
@Builder
@RequiredArgsConstructor
@AllArgsConstructor
public class PriceRequest {
  private String date;
  private Collection<AssetInput> assets;

  public static PriceRequestBuilder of(Asset asset) {
    Collection<AssetInput> assetInputs = new ArrayList<>();
    assetInputs.add(AssetInput.builder()
        .code(asset.getCode())
        .market(asset.getMarket().getCode())
        .resolvedAsset(asset)
        .build());

    return builder()
        .assets(assetInputs);
  }
}

File: jar-common/src/main/java/com/beancounter/common/contracts/PriceRequest.kt
package com.beancounter.common.contracts

import com.beancounter.common.input.AssetInput
import com.beancounter.common.model.Asset
import com.beancounter.common.utils.AssetUtils
import com.beancounter.common.utils.DateUtils.Companion.today

/**
 * Arguments by which prices on a date are located.
 */
data class PriceRequest(val date: String = today, val assets: Collection<AssetInput>) {

    companion object {
        @JvmStatic
        fun of(asset: Asset): PriceRequest {
            val assetInputs: MutableCollection<AssetInput> = ArrayList()
            assetInputs.add(AssetUtils.getAssetInput(asset))
            return PriceRequest(today, assetInputs)
        }

        @JvmStatic
        fun of(assetInput: AssetInput): PriceRequest {
            val assetInputs: MutableCollection<AssetInput> = ArrayList()
            assetInputs.add(assetInput)
            return PriceRequest(today, assetInputs)
        }
    }
}
